src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -0,0 +1,1395 @@
+
+    let batch_size = 2;
+    let timestamp_interval = Duration::from_secs(1);
+    let num_readers = 1;
+
+    // Make this artificially slow so we can eyeball the output.
+    let source_wait_time = Duration::from_secs(2);
+    let now_fn = SYSTEM_TIME.clone();
+
+    let source = SequentialSource::new(2, source_wait_time);
+
+    let persist = persist_client(
+        "file:///tmp/persistent_source_example/blob",
+        "sqlite:///tmp/persistent_source_example/sqlite-consensus",
+    )
+    .await?;
+
+    let bindings_id = "sa3150c18-f7a0-4062-b189-463531b9f36a".parse()?;
+    let data_id = "s1ba38f23-e28f-41fc-a92b-cddef58c48ce".parse()?;
+
+    let (bindings_write, bindings_read) = persist.open(bindings_id).await?;
+
+    let consensus = PersistConsensus::new(bindings_write, bindings_read);
+    let timestamper =
+        ConsensusTimestamper::new(now_fn.clone(), timestamp_interval.clone(), consensus);
+
+    let (data_write, mut data_read) = persist.open(data_id).await?;
+
+    // First, render one instance of the source.
+    let source1 = source.clone();
+    let source_pipeline1 = mz_ore::task::spawn(|| "source-1", async move {
+        if let Err(e) = render::spawn_source_pipeline(
+            "source-1".to_string(),
+            source1,
+            timestamper,
+            batch_size,
+            data_write,
+        )
+        .await
+        {
+            error!("error in source: {:?}", e);
+        }
+    });
+
+    let mut reader_pipelines = vec![];
+    for i in 0..num_readers {
+        let reader_name = format!("reader-{}", i);
+        let reader_name_clone = reader_name.clone();
+        let (_write, data_read) = persist.open::<String, (), Timestamp, _>(data_id).await?;
+        let pipeline = mz_ore::task::spawn(|| &reader_name_clone, async move {
+            let as_of = data_read.since().clone();
+            if let Err(e) = reader::spawn_reader_pipeline(reader_name, data_read, as_of).await {
+                error!("error in reader: {:?}", e);
+            }
+        });
+
+        reader_pipelines.push(pipeline);
+    }
+
+    // Make it explicit that we will never read from this, but only after creating the "real"
+    // reader. Otherwise, that real reader couldn't read because the global since would have
+    // advanced.
+    data_read.downgrade_since(Antichain::new()).await;
+
+    // After a while, start a second pipeline that reads from the same source and shares the
+    // timestamper.
+    tokio::time::sleep(Duration::from_secs(5)).await;
+
+    let (bindings_write, bindings_read) = persist.open(bindings_id).await?;
+    let consensus = PersistConsensus::new(bindings_write, bindings_read);
+    let timestamper =
+        ConsensusTimestamper::new(now_fn.clone(), timestamp_interval.clone(), consensus);
+
+    let (data_write, mut data_read) = persist.open(data_id).await?;
+    // Make it explicit that we will never read from this.
+    data_read.downgrade_since(Antichain::new()).await;
```

Comment:
is drop just as good for this? that wouldn't break if we switch downgrade_since to be shard-global

Reply:
Yes, `drop` is actually better.